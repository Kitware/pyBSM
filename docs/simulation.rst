##########
Simulation
##########

.. autosummary::
   :toctree: _implementations/simulation
   :template: custom-module-template.rst
   :recursive:

   pybsm.simulation.functional
   pybsm.simulation.ref_image
   pybsm.simulation.scenario
   pybsm.simulation.sensor
