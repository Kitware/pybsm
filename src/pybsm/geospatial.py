"""The Python Based Sensor Model (pyBSM) is a collection of electro-optical camera modeling functions.

Developed by the Air Force Research Laboratory, Sensors Directorate.

Author citation:
LeMaster, Daniel A.; Eismann, Michael T., "pyBSM: A Python package for modeling
imaging systems", Proc. SPIE 10204 (2017)

Distribution A.  Approved for public release.
Public release approval for version 0.0: 88ABW-2017-3101
Public release approval for version 0.1: 88ABW-2018-5226

Maintainer: Kitware, Inc. <nrtk@kitware.com>
"""

# standard library imports
import inspect
import os
import warnings

# 3rd party imports
import numpy as np

# new in version 0.2.  We filter warnings associated with calculations in the
# function circularApertureOTF.  These invalid values are caught as NaNs and
# appropriately replaced.
warnings.filterwarnings("ignore", r"invalid value encountered in arccos")
warnings.filterwarnings("ignore", r"invalid value encountered in sqrt")
warnings.filterwarnings("ignore", r"invalid value encountered in true_divide")
warnings.filterwarnings("ignore", r"divide by zero encountered in true_divide")

# find the current path (used to locate the atmosphere database)
# dir_path = os.path.dirname(os.path.abspath(__file__))
dir_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))

# define some useful physical constants
r_earth = 6378.164e3  # radius of the earth (m)


def altitude_along_slant_path(
    h_target: float,
    h_sensor: float,
    slant_range: float,
) -> tuple[np.ndarray, np.ndarray]:
    """Calculate the height above the curved earth at points along a path from the target to the sensor.

    Calculate the height above the curved earth at points along a path from the
    target (z_path=0) to the sensor (z_path.max()).  This is primarily useful for
    calculating the atmospheric coherence diameter, r0.

    :param h_target:
        height of the target above sea level (m)
    :param h_sensor:
        height of the sensor above sea level (m)
    :param slant_range:
        distance between the target and sensor (m)

    :return:
        z_path:
            array of samples along the path from the target (z_path = 0) to the
            sensor (m)
        h_path:
            height above the earth along a slantpath defined by z_path (m)

    :raises:
        ZeroDivisionError:
            if slant_range is 0

    """
    # this is simple law of cosines problem
    nadir = nadir_angle(h_target, h_sensor, slant_range)
    a = r_earth + h_sensor
    b = np.linspace(0.0, slant_range, 10000)  # arbitrary choice of 100,000 data points
    c = np.sqrt(a**2 + b**2 - 2 * a * b * np.cos(nadir))

    z_path = b
    h_path = (c - r_earth)[::-1]
    # It is correct to reverse the order of h_path.  The height of the target
    # above the earth (h_target) should be the first element in the array.
    # Effectively, this reversal just changes the location of the origin of
    # z_path to the target location.  It was just more convenient to use the
    # origin at the sensor in the law of cosines calculation.

    return (z_path, h_path)


def ground_sample_distance(ifov: float, slant_range: float) -> float:
    """IBSM Equation 3-62, the ground sample distance.

    The ground sample distance, i.e. the footprint of a single detector in object space.

    :param ifov:
        instantaneous field-of-view of a detector (radians)
    :param slant_range:
        slant range to target (m)

    :return:
        gsd:
            ground sample distance (m)
    """
    return slant_range * ifov


def nadir_angle(h_target: float, h_sensor: float, slant_range: float) -> float:
    """Work through the law of cosines to calculate the sensor nadir angle above a circular earth.

    Work through the law of cosines to calculate the sensor nadir angle above a
    circular earth (i.e. angle between looking straight down (nadir = 0) and looking
    along the slant path).

    :param h_target:
        height of the target above sea level (m)
    :param h_sensor:
        height of the sensor above sea level (m)
    :param slant_range:
        distance between the target and sensor (m)

    :return:
        nadir:
            the sensor nadir angle (rad)

    :raises:
        ZeroDivisionError:
            if slant_range is 0
    """
    a = r_earth + h_sensor
    b = slant_range
    c = r_earth + h_target

    return np.arccos(-1.0 * (c**2.0 - a**2.0 - b**2.0) / (2.0 * a * b))


def curved_earth_slant_range(
    h_target: float,
    h_sensor: float,
    ground_range: float,
) -> float:
    """Returns the slant range from target to sensor above a curved (circular) Earth.

    :param h_target:
        height of the target above sea level (m)
    :param h_sensor:
        height of the sensor above sea level (m)
    :param ground_range:
        distance between the target and sensor on the ground (m)

    :return:
        slant_range:
            distance between the target and sensor (m)
    """
    a = r_earth + h_sensor
    c = r_earth + h_target
    theta = ground_range / r_earth  # exact arc length angle (easy to derive)

    return np.sqrt(c**2.0 + a**2.0 - 2.0 * a * c * np.cos(theta))
