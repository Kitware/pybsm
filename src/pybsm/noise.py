"""The Python Based Sensor Model (pyBSM) is a collection of electro-optical camera modeling functions.

Developed by the Air Force Research Laboratory, Sensors Directorate.

Author citation:
LeMaster, Daniel A.; Eismann, Michael T., "pyBSM: A Python package for modeling
imaging systems", Proc. SPIE 10204 (2017)

Distribution A.  Approved for public release.
Public release approval for version 0.0: 88ABW-2017-3101
Public release approval for version 0.1: 88ABW-2018-5226

Maintainer: Kitware, Inc. <nrtk@kitware.com>
"""

# standard library imports
import inspect
import os
import warnings

# 3rd party imports
import numpy as np

# new in version 0.2.  We filter warnings associated with calculations in the
# function circularApertureOTF.  These invalid values are caught as NaNs and
# appropriately replaced.
warnings.filterwarnings("ignore", r"invalid value encountered in arccos")
warnings.filterwarnings("ignore", r"invalid value encountered in sqrt")
warnings.filterwarnings("ignore", r"invalid value encountered in true_divide")
warnings.filterwarnings("ignore", r"divide by zero encountered in true_divide")

# find the current path (used to locate the atmosphere database)
# dir_path = os.path.dirname(os.path.abspath(__file__))
dir_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))


def noise_gain(kernel: np.ndarray) -> float:
    """Noise Gain is the GIQE term representing increase in noise due to image sharpening.

    The definition is not included in the IBSM manual.  This version comes from
    Leachtenauer et al., "General Image-Quality Equation: GIQE" APPLIED OPTICS
    Vol. 36, No. 32 10 November 1997.

    :param kernel:
         the 2D image sharpening kernel; note that
         the kernel is assumed to sum to one

    :return:
        ng:
            noise gain (unitless)

    :raises:
        ValueError:
            if values in kernel do not sum to 1
    """
    if np.sum(kernel) != 1.0:
        raise ValueError("Kernel does not sum to 1")

    return np.sqrt(np.sum(np.sum(kernel**2)))


def quantization_noise(pe_range: float, bit_depth: float) -> float:
    """Effective noise contribution from the number of photoelectrons quantized.

    Effective noise contribution from the number of photoelectrons quantized by a single count of the analog
    to digital converter.Quantization noise is buried in the definition of signal-to-noise in IBSM equation 3-47.

    :param pe_range:
        the difference between the maximum and minimum number of photoelectrons
        that may be sampled by the readout electronics (e-)
    :param bit_depth:
        number of bits in the analog to digital converter (unitless)

    :return:
        sigma_q :
            quantization noise given as a photoelectron standard deviation (e-)

    :WARNING:
        output can be nan if pe_range is 0
        output can be inf if bit_depth is 0
    """
    return pe_range / (np.sqrt(12) * (2.0**bit_depth - 1.0))
