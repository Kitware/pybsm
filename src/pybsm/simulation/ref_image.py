"""The Python Based Sensor Model (pyBSM) is a collection of electro-optical camera modeling functions.

Developed by the Air Force Research Laboratory, Sensors Directorate.

Author citation:
LeMaster, Daniel A.; Eismann, Michael T., "pyBSM: A Python package for modeling
imaging systems", Proc. SPIE 10204 (2017)

Distribution A.  Approved for public release.
Public release approval for version 0.0: 88ABW-2017-3101
Public release approval for version 0.1: 88ABW-2018-5226

Maintainer: Kitware, Inc. <nrtk@kitware.com>
"""

# 3rd party imports
from typing import Optional

import matplotlib.pyplot as plt
import numpy as np

from .scenario import Scenario

# local imports
from .sensor import Sensor


class RefImage:
    """Reference image.

    Attributes:
        img (np.ndarray):
            The reference image data.
        gsd (float):
            The ground sampling distance in meters.
        pix_values (np.ndarray):
            The pixel count values for interpolation.
        refl_values (np.ndarray):
            The reflectance values associated with 'pix_values'.
        name (str):
            The name of the reference image.
    """

    def __init__(
        self,
        img: np.ndarray,
        gsd: float,
        pix_values: Optional[np.ndarray] = None,
        refl_values: Optional[np.ndarray] = None,
        name: str = "ref_image",
    ) -> None:
        """
        Initializes a reference image.

        :param img: Reference image.
        :param gsd:
            Spatial sampling for 'img' in meters. Each pixel in 'img' is assumed to
            capture a 'gsd' x 'gsd' square of some world surface. We assume the
            sampling is isotropic (x and y sampling are identical) and uniform
            across the whole field of view. This is generally a valid assumption
            for remote sensing imagery.
        :param pix_values:
            Pixel count values within 'img' that should be associated with the
            corresponding reflectance values in 'refl_values' by linear
            interpolation. This is used to convert raw image values into an assumed
            spectral reflectance of the scene being viewed.
        :param refl_values:
            Reflectance values associated with the corresponding pixel count values
            in 'pix_values' used to convert raw image values into an assumed
            spectral reflectance of the scene being viewed.
        :param name: Name of the image.

        :raises: ValueError if pix_values is provided, but refl_values is missing
        """
        self.img = img
        self.gsd = gsd
        self.name = name

        if pix_values is None:
            pix_values = np.array(
                [
                    np.percentile(img.ravel(), 0.2),
                    np.percentile(img.ravel(), 99.8),
                ],
            )
            refl_values = np.array([0.05, 0.95])
        else:
            if refl_values is None:
                raise ValueError("If 'pix_values' is provided, 'refl_values' must be as well.")

        self.pix_values = pix_values
        self.refl_values = refl_values

    def estimate_capture_parameters(self, altitude: float = 2000000) -> tuple[Sensor, Scenario]:
        """Estimate the scenario and sensor parameters that are consistent with this image.

        This provides a no-degradation baseline from which to alter parameters
        to explore further degradation.
        """
        # Let's assume the full visible spectrum.
        opt_trans_wavelengths = np.array([380, 700]) * 1.0e-9  # m

        scenario = Scenario(
            self.name,
            1,
            altitude,
            ground_range=0,
            aircraft_speed=0,
            ha_wind_speed=0,
            cn2_at_1m=0,
        )

        # Guess at a relatively large pixel pitch, which should have a large
        # well depth.
        p = 20e-6  # m

        # Calculate the focal length (m) that matches the GSD for the
        # prescribed altitude and pixel pitch.
        f = altitude * p / self.gsd  # m

        # Instantenous field of view (iFOV), the angular extent of the world
        # covered by one pixel (radians).
        ifov = 2 * np.arctan(p / 2 / f)

        # We are assuming a circular aperture without obscuration. The
        # diffraction limited angular resolution (where on Airy disk sits in
        # the first ring of another Airy disk) is 1.22*lambda/D. But, let's use
        # a coefficient of 4 for safety.
        D = 4 * np.median(opt_trans_wavelengths) / ifov  # noqa: N806

        sensor = Sensor(self.name, D, f, p, opt_trans_wavelengths)
        return sensor, scenario

    def show(self) -> None:
        """Plots this image."""
        h, w = self.img.shape[:2]
        plt.imshow(
            self.img,
            extent=(
                -w / 2 * self.gsd,
                w / 2 * self.gsd,
                -h / 2 * self.gsd,
                h / 2 * self.gsd,
            ),
        )
        plt.xlabel("X-Position (m)", fontsize=24)
        plt.ylabel("Y-Position (m)", fontsize=24)
        plt.tight_layout()
