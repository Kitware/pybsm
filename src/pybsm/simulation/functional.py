"""The Python Based Sensor Model (pyBSM) is a collection of electro-optical camera modeling functions.

Developed by the Air Force Research Laboratory, Sensors Directorate.

Author citation:
LeMaster, Daniel A.; Eismann, Michael T., "pyBSM: A Python package for modeling
imaging systems", Proc. SPIE 10204 (2017)

Distribution A.  Approved for public release.
Public release approval for version 0.0: 88ABW-2017-3101
Public release approval for version 0.1: 88ABW-2018-5226

Maintainer: Kitware, Inc. <nrtk@kitware.com>
"""

# standard library imports
import inspect
import logging
import os
import warnings
from typing import Optional, Union

# 3rd party imports
import numpy as np
from scipy import interpolate

# local imports
import pybsm.otf as otf
import pybsm.radiance as radiance
from pybsm import noise

from .ref_image import RefImage
from .scenario import Scenario
from .sensor import Sensor

# new in version 0.2.  We filter warnings associated with calculations in the
# function circularApertureOTF.  These invalid values are caught as NaNs and
# appropriately replaced.
warnings.filterwarnings("ignore", r"invalid value encountered in arccos")
warnings.filterwarnings("ignore", r"invalid value encountered in sqrt")
warnings.filterwarnings("ignore", r"invalid value encountered in true_divide")
warnings.filterwarnings("ignore", r"divide by zero encountered in true_divide")

# find the current path (used to locate the atmosphere database)
# dir_path = os.path.dirname(os.path.abspath(__file__))
dir_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))


def instantaneous_FOV(w: int, f: int) -> float:  # noqa: N802
    """The instantaneous field of view; i.e., the angular footprint of a single detector in object space.

    :param w:
        detector size (width) in the x and y directions (m)
    :type w: int
    :param f:
        focal length (m)
    :type f: int

    :return:
        ifov:
            detector instantaneous field-of-view (radians)

    :raises:
        ZeroDivisionError:
            error raised when f is 0

    """
    return w / f


def wiener_filter(
    otf: np.ndarray,
    noise_to_signal_power_spectrum: float,
) -> np.ndarray:
    """An image restoration transfer function based on the Wiener Filter.

    See from Gonzalex and Woods, "Digital Image Processing," 3 ed.  Note that the
    filter is not normalized so that WF = 1 at 0 spatial frequency.  This is
    easily fixed for the case where noise_to_signal_power_spectrum is a
    scalar: (1.0+noisetosignalPS)*WF = 1 at 0 spatial frequency.
    This is noteworthy because, without normalization of some kind, system
    MTF * WF is not a proper MTF. Also, for any frequency space sharpening
    kernel, be sure that sharpening does not extend past the Nyquist frequency,
    or, if so, ensure that the filter response wraps around Nyquist
    appropriately.

    :param otf:
        system optical transfer function
    :type otf: list
    :param noiseTosignalPS:
        ratio of the noise power spectrum to the signal power spectrum; this
        may be a function of spatial frequency (same size as otf) or a scalar
    :type noiseTosignalPS: float

    :return:
        WF:
            frequency space representation of the wiener_filter
    """
    return np.conj(otf) / (np.abs(otf) ** 2 + noise_to_signal_power_spectrum)


def img_to_reflectance(
    img: np.ndarray,
    pix_values: np.ndarray,
    refl_values: np.ndarray,
) -> np.ndarray:
    """Maps pixel values to reflectance values with linear interpolation between points.

    Pixel values that map below zero reflectance or above unity reflectance are truncated.
    Implicitly, reflectance is contrast across the camera bandpass.

    :param img:
        the image that will be transformed into reflectance (counts)
    :type img: np.array
    :param pix_values:
        array of values in img that map to a known reflectance (counts)
    :type pix_values: np.array
    :param refl_values:
        array of reflectances that correspond to pix_values (unitless)
    :type refl_values: np.array

    :return:
        ref_img:
            the image in reflectance space

    :raises:
        ValueError:
            if img, pix_values, or refl_values have a length < 2

    :WARNING:
        Output can be nan if all input arrays have the same values.
    """
    f = interpolate.interp1d(
        pix_values,
        refl_values,
        fill_value="extrapolate",
        assume_sorted=0,
    )
    ref_img = f(img)
    ref_img[ref_img > 1.0] = 1.0
    ref_img[ref_img < 0.0] = 0.0

    return ref_img


def simulate_image(
    ref_img: RefImage,
    sensor: Sensor,
    scenario: Scenario,
    rng: Optional[Union[np.random.Generator, int]] = 1,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Simulates radiometrically accurate imagery collected through a sensor.

    We start with a notionally ideal reference image 'img', which captures a
    view of the world of which we would like to simulate a degraded view that
    would be collected from another imaging system. Our framework for image
    simulation requires that this reference image is of higher quality and
    ideally higher resolution than the view we would like to simulate as we can
    only model further degradation of image quality.


    :type ref_img: RefImage
    :param ref_img:
        Reference image to use as the source view of the world to be emulated
        by the virtual camera defined by 'sensor'.
    :type sensor: Sensor
    :param sensor:
        Virtual sensor definition.
    :type scenario: Scenario
    :param scenario:
        Specification of the deployment of the virtual sensor within the world
        relative to the target.
    :type rng: np.random.Generator or int
    :param rng:


    :return:
        true_img:
            Numpy float64 array; the true image in units of photoelectrons
        blur_img:
            Numpy float64 array; the image after blurring and resampling is applied to true_img
            (still units of photoelectrons)
        noisy_img:
            Numpy float64 array; the blur image with photon (Poisson) noise and gaussian noise
            applied (still units of photoelectrons)

    :WARNING:
        imggsd must be small enough to properly sample the blur kernel. As a
        guide, if the image system transfer function goes to zero at angular
        spatial frequency, coff, then the sampling requirement will be readily
        met if imggsd <= rng/(4*coff). In practice this is easily done by
        upsampling imgin.

    :raises: ValueError if cutoff Frequency matrix u_rng is not monotonically
             increasing
    """
    pybsm_rng = np.random.default_rng(rng)

    # integration time (s)
    int_time = sensor.int_time

    (
        ref,
        pe,
        spectral_weights,
    ) = radiance.reflectance_to_photoelectrons(scenario.atm, sensor, int_time)

    wavelengths = spectral_weights[0]
    weights = spectral_weights[1]

    slant_range = np.sqrt(scenario.altitude**2 + scenario.ground_range**2)

    # cut down the wavelength range to only the regions of interest
    mtf_wavelengths = wavelengths[weights > 0.0]
    mtf_weights = weights[weights > 0.0]

    # Assume if nothing else cuts us off first, diffraction will set the limit
    # for spatial frequency that the imaging system can resolve (1/rad).
    cutoff_frequency = sensor.D / np.min(mtf_wavelengths)

    u_rng = np.linspace(-1.0, 1.0, 1501) * cutoff_frequency
    v_rng = np.linspace(1.0, -1.0, 1501) * cutoff_frequency

    # meshgrid of spatial frequencies out to the optics cutoff
    uu, vv = np.meshgrid(u_rng, v_rng)

    system_otf = otf.common_OTFs(
        sensor,
        scenario,
        uu,
        vv,
        mtf_wavelengths,
        mtf_weights,
        slant_range,
        int_time,
    ).system_OTF

    df = (abs(u_rng[1] - u_rng[0]) + abs(v_rng[0] - v_rng[1])) / 2  # noqa: PD901

    if df <= 0:
        raise ValueError("Cutoff frequency values must be increasing.")

    ifov = (sensor.p_x + sensor.p_y) / 2 / sensor.f

    # Standard deviation of additive Gaussian noise (e.g. read noise,
    # quantization). Should be the RSS value if multiple terms are combined.
    # This should not include photon noise.
    quantization_noise = noise.quantization_noise(sensor.max_n, sensor.bit_depth)
    g_noise = np.sqrt(quantization_noise**2.0 + sensor.read_noise**2.0)

    # Convert to reference image into a floating point reflectance image.
    reflectance_img = img_to_reflectance(
        ref_img.img,
        ref_img.pix_values,
        ref_img.refl_values,
    )

    # Convert the reflectance image to photoelectrons.
    f = interpolate.interp1d(ref, pe)
    true_img = f(reflectance_img)

    # blur and resample the image
    blur_img, _ = otf.apply_otf_to_image(
        true_img,
        ref_img.gsd,
        slant_range,
        system_otf,
        df,
        ifov,
    )

    # add photon noise (all sources) and dark current noise
    poisson_noisy_img = pybsm_rng.poisson(lam=blur_img)
    # add any noise from Gaussian sources, e.g. read_noise, quantizaiton
    noisy_img = pybsm_rng.normal(poisson_noisy_img, g_noise)

    if noisy_img.shape[0] > ref_img.img.shape[0]:
        logging.warn(
            "The simulated image has oversampled the reference image!  This result should not be trusted!!",
        )

    return true_img, blur_img, noisy_img


def stretch_contrast_convert_8bit(
    img: np.ndarray,
    perc: Optional[list[float]] = None,
) -> np.ndarray:
    """
    Adjusts the contrast of an image and converts it to an 8-bit format.

    This function stretches the contrast of the input image based on specified
    percentile values. The pixel values are normalized to the 8-bit range (0-255)
    and any values outside this range are clipped.

    :param img:
        The original image.
    :param perc:
        A list of two percentile values (0-100) to determine the contrast stretch.
        The first value corresponds to the lower percentile and the second to the
        upper percentile. If not provided, defaults to [0.1, 99.9].

    :return:
        A numpy array of the same shape as `img`, with pixel values converted
        to 8-bit unsigned integers (0-255) after contrast stretching.

    Example:
        >>> import numpy as np
        >>> img = np.random.rand(100, 100) * 65535  # Example 16-bit image
        >>> stretched_img = stretch_contrast_convert_8bit(img)
    """
    if perc is None:
        perc = [0.1, 99.9]
    img = img.astype(float)
    img = img - np.percentile(img.ravel(), perc[0])
    img = img / (np.percentile(img.ravel(), perc[1]) / 255)
    img = np.clip(img, 0, 255)
    return np.round(img).astype(np.uint8)
