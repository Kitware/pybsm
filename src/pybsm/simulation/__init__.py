from .sensor import Sensor  # noqa F401
from .scenario import Scenario  # noqa F401
from .ref_image import RefImage  # noqa F401
from .functional import *  # noqa F401,F403
from .image_simulator import *  # noqa F401,F403
