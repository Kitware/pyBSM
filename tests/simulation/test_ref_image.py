import unittest.mock as mock
from contextlib import nullcontext as does_not_raise
from pathlib import Path
from typing import ContextManager

import matplotlib.pyplot as plt
import numpy as np
import pytest

from pybsm.simulation import RefImage, Scenario, Sensor

BASE_PATH = Path(__file__).parent.parent.parent
IMAGE_PATH = BASE_PATH / 'examples' / 'data' / 'M-41 Walker Bulldog (USA) width 319cm height 272cm.tiff'


class TestRefImage:
    @pytest.mark.parametrize(
        ("pix_values", "refl_values", "expectation"),
        [
            (None, None, does_not_raise()),
            (np.array([0.05, 0.95]), None,
             pytest.raises(ValueError, match="If 'pix_values' is provided, 'refl_values' must be as well.")),
            (None, np.array([0.05, 0.95]), does_not_raise()),
            (np.array([0.05, 0.95]), np.array([0.1, 0.9]), does_not_raise())

        ]
    )
    def test_ref_image_init(self,
                            pix_values: np.ndarray,
                            refl_values: np.ndarray,
                            expectation: ContextManager) -> None:
        img = plt.imread(IMAGE_PATH)
        gsd = (3.19 / 160.0)
        with expectation:
            RefImage(img, gsd, pix_values, refl_values)

    @pytest.mark.parametrize(
            ("altitude", "expected_sensor", "expected_scenario"),
            [(1000,
             Sensor("ref_image",
                    0.10833855799731913,
                    1.0031347962382444,
                    20e-6,
                    np.array([3.8e-07, 7.0e-07])),
             Scenario("ref_image",
                      1,
                      1000,
                      ground_range=0,
                      aircraft_speed=0,
                      ha_wind_speed=0,
                      cn2_at_1m=0)),
             (10000,
             Sensor("ref_image",
                    1.0833855799376628,
                    10.031347962382446,
                    20e-6,
                    np.array([3.8e-07, 7.0e-07])),
             Scenario("ref_image",
                      1,
                      10000,
                      ground_range=0,
                      aircraft_speed=0,
                      ha_wind_speed=0,
                      cn2_at_1m=0))]
    )
    def test_estimate_capture_parameters(self,
                                         altitude: float,
                                         expected_sensor: Sensor,
                                         expected_scenario: Scenario) -> None:
        img = plt.imread(IMAGE_PATH)
        gsd = (3.19 / 160.0)
        ref_image = RefImage(img, gsd)
        (sensor, scenario) = ref_image.estimate_capture_parameters(altitude)
        assert sensor.name == expected_sensor.name
        assert sensor.D == expected_sensor.D
        assert sensor.f == expected_sensor.f
        assert sensor.p_x == expected_sensor.p_x
        assert np.equal(sensor.opt_trans_wavelengths.all(),
                        expected_sensor.opt_trans_wavelengths.all())

        assert scenario.name == expected_scenario.name
        assert scenario.ihaze == expected_scenario.ihaze
        assert scenario.altitude == expected_scenario.altitude
        assert scenario.ground_range == expected_scenario.ground_range
        assert scenario.aircraft_speed == expected_scenario.aircraft_speed
        assert scenario.ha_wind_speed == expected_scenario.ha_wind_speed
        assert scenario.cn2_at_1m == expected_scenario.cn2_at_1m

    @mock.patch("matplotlib.pyplot.imshow")
    @mock.patch("matplotlib.pyplot.xlabel")
    @mock.patch("matplotlib.pyplot.ylabel")
    @mock.patch("matplotlib.pyplot.tight_layout")
    def test_show_method(self,
                         mock_tight_layout: mock.MagicMock,
                         mock_ylabel: mock.MagicMock,
                         mock_xlabel: mock.MagicMock,
                         mock_imshow: mock.MagicMock) -> None:
        img = plt.imread(IMAGE_PATH)
        gsd = (3.19 / 160.0)
        # Create the instance of TestClass
        obj = RefImage(img, gsd)

        # Call the show method
        obj.show()

        # Check if the imshow was called with correct extent calculation
        h, w = img.shape[:2]
        expected_extent = [
            -w / 2 * gsd,
            w / 2 * gsd,
            -h / 2 * gsd,
            h / 2 * gsd,
        ]
        assert mock_imshow.call_count == 1
        call_args, call_kwargs = mock_imshow.call_args

        # Validate the image argument using numpy.testing
        np.testing.assert_array_equal(call_args[0], img)

        # Validate the extent argument
        assert call_kwargs['extent'] == expected_extent

        # Verify xlabel was called with correct parameters
        mock_xlabel.assert_called_once_with("X-Position (m)", fontsize=24)

        # Verify ylabel was called with correct parameters
        mock_ylabel.assert_called_once_with("Y-Position (m)", fontsize=24)

        # Verify tight_layout was called once
        mock_tight_layout.assert_called_once()
