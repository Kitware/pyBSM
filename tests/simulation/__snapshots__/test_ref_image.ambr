# serializer version: 1
# name: TestRefImage.test_estimate_capture_parameters[10000]
  tuple(
    Sensor(
      D=1.0833855799376628,
      L_x=1.0833855799376628,
      L_y=1.0833855799376628,
      bit_depth=100.0,
      cold_filter_emissivity=0.0,
      cold_filter_temperature=70.0,
      cold_filter_transmission=1.0,
      cold_shield_temperature=70.0,
      da_x=0.0,
      da_y=0.0,
      dark_current=0,
      eta=0.0,
      f=10.031347962382446,
      filter_kernel=array([1]),
      frame_stacks=1,
      int_time=1,
      max_n=100000000,
      max_well_fill=1.0,
      n_tdi=1.0,
      name='ref_image',
      opt_trans_wavelengths=array([3.8e-07, 7.0e-07]),
      optics_emissivity=0.0,
      optics_temperature=270.0,
      optics_transmission=array([1., 1.]),
      other_irradiance=0.0,
      other_noise=array([0]),
      p_x=2e-05,
      p_y=2e-05,
      pv=0.0,
      pv_wavelength=6.33e-07,
      qe=array([1., 1.]),
      qe_wavelengths=array([3.8e-07, 7.0e-07]),
      read_noise=0,
      s_x=0.0,
      s_y=0.0,
      w_x=2e-05,
      w_y=2e-05,
    ),
    Scenario(
      aircraft_speed=0,
      altitude=10000,
      atm=array([[3.00000000e-07, 1.61430404e-01, 0.00000000e+00, 0.00000000e+00,
          2.96549975e+06, 1.98660000e+05],
         [3.10000000e-07, 2.37839296e-01, 0.00000000e+00, 0.00000000e+00,
          1.22560000e+07, 9.74010000e+05],
         [3.20000000e-07, 2.90373504e-01, 0.00000000e+00, 0.00000000e+00,
          2.76150000e+07, 2.60459975e+06],
         ...,
         [1.39800000e-05, 1.47225996e-02, 4.15140000e+06, 9.23430000e+04,
          2.30259988e-02, 1.48580000e+04],
         [1.39900000e-05, 6.33369992e-03, 4.04519975e+06, 3.96770000e+04,
          1.19440006e-02, 6.45859961e+03],
         [1.40000000e-05, 3.90179991e-03, 4.15619975e+06, 2.44080000e+04,
          1.04599996e-02, 4.00289990e+03]]),
      background_reflectance=0.07,
      background_temperature=293,
      cn2_at_1m=0,
      ground_range=0,
      ha_wind_speed=0,
      ihaze=1,
      name='ref_image',
      target_reflectance=0.15,
      target_temperature=295,
    ),
  )
# ---
# name: TestRefImage.test_estimate_capture_parameters[1000]
  tuple(
    Sensor(
      D=0.10833855799731913,
      L_x=0.10833855799731913,
      L_y=0.10833855799731913,
      bit_depth=100.0,
      cold_filter_emissivity=0.0,
      cold_filter_temperature=70.0,
      cold_filter_transmission=1.0,
      cold_shield_temperature=70.0,
      da_x=0.0,
      da_y=0.0,
      dark_current=0,
      eta=0.0,
      f=1.0031347962382444,
      filter_kernel=array([1]),
      frame_stacks=1,
      int_time=1,
      max_n=100000000,
      max_well_fill=1.0,
      n_tdi=1.0,
      name='ref_image',
      opt_trans_wavelengths=array([3.8e-07, 7.0e-07]),
      optics_emissivity=0.0,
      optics_temperature=270.0,
      optics_transmission=array([1., 1.]),
      other_irradiance=0.0,
      other_noise=array([0]),
      p_x=2e-05,
      p_y=2e-05,
      pv=0.0,
      pv_wavelength=6.33e-07,
      qe=array([1., 1.]),
      qe_wavelengths=array([3.8e-07, 7.0e-07]),
      read_noise=0,
      s_x=0.0,
      s_y=0.0,
      w_x=2e-05,
      w_y=2e-05,
    ),
    Scenario(
      aircraft_speed=0,
      altitude=1000,
      atm=array([[3.00000000e-07, 6.70006573e-01, 0.00000000e+00, 0.00000000e+00,
          4.59480000e+05, 5.58120000e+05],
         [3.10000000e-07, 7.08292782e-01, 0.00000000e+00, 0.00000000e+00,
          1.93110000e+06, 2.50940000e+06],
         [3.20000000e-07, 7.29579687e-01, 0.00000000e+00, 0.00000000e+00,
          4.43270000e+06, 6.22370000e+06],
         ...,
         [1.39800000e-05, 1.38834193e-01, 5.77420000e+06, 8.70480000e+05,
          1.35859996e-02, 1.40620000e+05],
         [1.39900000e-05, 8.74299034e-02, 6.10450000e+06, 5.47540000e+05,
          5.68979979e-03, 8.93820000e+04],
         [1.40000000e-05, 6.36345968e-02, 6.25160000e+06, 3.98050000e+05,
          3.34699987e-03, 6.52830000e+04]]),
      background_reflectance=0.07,
      background_temperature=293,
      cn2_at_1m=0,
      ground_range=0,
      ha_wind_speed=0,
      ihaze=1,
      name='ref_image',
      target_reflectance=0.15,
      target_temperature=295,
    ),
  )
# ---
# name: TestRefImage.test_show_method
  tuple(
    -5.104,
    5.104,
    -5.104,
    5.104,
  )
# ---
